File: python/heap.py
# Implementation of heap algorithms

from heapq import heappush, heappop, heapify

def sort_heap(seq):
    h = []
    for v in seq :
        heappush(h, v)
    return [heappop(h) for i in range(len(h))]

def is_heap(seq):
    if len(seq) == 0 : return True
    for i in range(1, len(seq)) :
        if seq[i] < seq[int((i - 1)/ 2)] :
            return False
    return True

def is_heap_until(seq):
    if len(seq) == 0 : return 0
    for i in range(1, len(seq)) :
        if seq[i] < seq[int((i - 1)/ 2)] :
            return i
    return len(seq)

def push_heap(seq, value):
    seq.append(value)
    heapify(seq)

def pop_heap(seq):
    tmp = seq.pop(0)
    heapify(seq)
    return tmp

# the heap made by heapq.heapify is a min-heap, so we construct a min-heap here
def make_heap(seq):
    heap_size = len(seq)
    for cur in range(int(heap_size / 2) - 1, -1, -1) :
        while True :
            left = 2 * cur + 1
            if left >= heap_size : break
            if seq[left] < seq[cur] : smallest = left
            else : smallest = cur
            right = left + 1
            if right < heap_size and seq[right] < seq[smallest] :
                smallest = right
            if seq[smallest] != seq[cur] :
                seq[cur], seq[smallest] = seq[smallest], seq[cur]
                cur = smallest
            else : break

File: ruby/heap.rb
# Implementation of heap algorithms

module Heap

def is_heap(seq)
  (1...seq.size).each{|i| return false if seq[i] < seq[((i-1)/2).to_i]}
  true
end

def is_heap_until(seq)
  (1...seq.size).each{|i| return i if seq[i] < seq[((i-1)/2).to_i]}
  seq.size
end

def push_heap(seq, value)
  seq.push value
  make_heap seq
end

def pop_heap(seq)
  res = seq.shift
  make_heap seq
  res
end

def sort_heap(seq)
  make_heap seq
  Array.new(seq.size){pop_heap(seq)}
end

# make a min heap
def make_heap(seq)
  (0...(seq.size/2).to_i).reverse_each do |i|
    loop do
      left = 2 * i + 1
      break if left >= seq.size
      smallest = (seq[left] < seq[i]) ? left : i
      right = left + 1
      smallest = right if right < seq.size && seq[right] < seq[smallest]
      break if seq[smallest] == seq[i]
      seq[i], seq[smallest] = seq[smallest], seq[i]
      i = smallest
    end
  end
end

end
